### PR TITLE
language-agnostic normalization steps

### PR DESCRIPTION
**Pull request by zupon**
This PR includes code to implement the following language-agnostic normalization steps:
1. Removing extra whitespace characters
2. Separating initial and final punctuation from words
3. Deleting freestanding punctuation
4. Lowercasing the sentence, and
5. Filtering out tokens containing characters not in the language

This PR also adds config files for the languages we've chosen as well as the start of a preprocessing script (`preprocess.py`) to handle the raw data from our different data sources.

### CODE REVIEW

File: normalizer.py
Comment (crewee) [inline normalizer.py]: Can we rename these to not be "Universal Dependencies" specific?  Maybe something like "DATA_PATH" and "DATA_TEXT".

Comment (crewee) [reply normalizer.py]: We also want to keep this language netural here, so this should perhaps be a flag that takes a command line argument filepath

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: language agnostic?  We should use a sample language only in the test file and keep this all language neutral.

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: For this purpose, this is fine, since we're not dealing with other numerical systems in these particular languages that we're working with, but as a note, this should be customizable in the language specific configurations as well, so that we can include locale specific digit sets :).

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Instead of naming the cdrewrites "DO_", let's name them what they're doing, like "REMOVE_EXTRA_WHITESPACE", then rename the transducer to something more simple?

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Since you're not reusing this, you can just put transducer(SPACE, "") in the cdrewrite itself!

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Let's name this function "pass_only_valid_tokens", so that it's clear that it checks words and not sentences

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Also, could you beef up the docstring comment to explain that we're checking individual tokens and not full sentences?

Additionally, for the docstrings, can you add "returns:" and "args:"?  Here's an example: https://google.github.io/styleguide/pyguide.html#244-decision

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: This will currently reject any email addresses, times, websites, words that are number/letter combos like "mp3", etc.  This works for this inital pull request, so no need to make changes before we merge this.  Perhaps a good next step for a follow up merge could be to define what we want to accept by way of these not-a-word tokens.  You could define acceptors for these like:

email_address = GRAPHEMES.plus + "@" + GRAPHEMES.plus + closure("." + GRAPHEMES.plus, 1, 2)

This would accept the following:
crewe@google.com
crewe@google.co.uk
etc.

Then you put them all together by saying something like:
def verbalizable(self):
  return email_address @ web_address

Then you include "self.verbalizable" in your token acceptor for passing valid tokens.

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Let's change the name of this to "INSERT_SPACE" to be more clear about what this is doing

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: This could just be called DELETE_PUNCTUATION, then you can call DO_DELETE_FREESTANDING_PUNCTUATION just DELETE_FREESTANDING_PUNCTUATION. 

The "freestanding" aspect is actually defined in the cdrewrite in the context that you give it, so that's why the naming suggestion :)

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py:166]: We may want to be careful about saying delete any punctuation that's at the beginning of the string, as it will remove word initial apostrophes if apostrophe is defined in PUNCTUATION.

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Let's just call this "normalizer" to simplify it :)

File: normalizer_test.py
Comment (crewee) [inline normalizer_test.py]: "TestNormalizer"?

File: normalizer_test.py
Comment (crewee) [inline normalizer_test.py:2]: The unit test looks great!  You could try using the parameterized tests that Sandy had previously mentioned!  Here's an example:
https://pypi.org/project/parameterized/

**Review by crewee — CHANGES_REQUESTED** [review]
Great work Andrew!!  It looks like a lot of comments, but mostly style guide related stuff.  Content looks good and an exciting first merge!

File: normalizer.py
Comment (zupon) [reply normalizer.py]: That makes sense. One complication is that right now it's calling the `process_ud_data` function from `preprocess.py`, which depends on the format of the UD data but not the other data sources. This should be easily fixed by making a generic `preprocess` function called here and adding different cases in `preprocess.py` that will vary depending on the data source.

Since this PR doesn't include `preprocess.py`, I'll change it here in `normalizer.py` for now and make sure to add those changes to the other file.

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Agreed!

File: normalizer_test.py
Comment (zupon) [reply normalizer_test.py]: Makes sense! I had left it unchanged from the starter project, but it should be more relevant to these actual tests.

File: normalizer_test.py
Comment (zupon) [reply normalizer_test.py:2]: For this and the starter project I used `subtest` from the `unittest` module (https://docs.python.org/3/library/unittest.html), which I _think_ does the same thing. For each of the tests (e.g. `test_remove_extra_whitespace`), if one of the test_cases fails, the whole test doesn't break; instead it keeps running the rest of the test_cases and reports the individual cases that fail rather than just saying the whole test failed.

If `parameterized` is what you actually use / what you are more familiar with though, I can definitely try to switch the tests over to that. (Plus it will let me get more experience with other Python modules).

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: This can easily go into the language-specific configs and called here as `NUMBERS = LANGUAGE.NUMBERS` like the graphemes, initial, and final punctuation.

File: normalizer_test.py
Comment (crewee) [reply normalizer_test.py:2]: Ah cool!  I mean, I don't think anyone's obsessed with the "parameterized" tests, so if this is a good alternative SGTM!  I think the parameterized tests can also end up taking a long time :).

File: normalizer_lib.py
Comment (crewee) [reply normalizer_lib.py]: SGTM thanks!

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Good point! And then it's easy enough to rename DO_REMOVE_EXTRA_WHITESPACE to just REMOVE_EXTRA_WHITESPACE

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Agreed!

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Sounds good. I tweaked the one-sentence description and added `args` and `returns` thus:

```
"""Replaces invalid tokens in a sentence with <UNK>, keeps the rest.

Args:
    string: A line from a corpus.

Returns:
    The line, but where invalid tokens have been replaced by <UNK>.
"""
```

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Done

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Done

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Makes sense! I think I just carried over the naming from the starter project, where there was a SOFT_MUTATION string_map and a DO_SOFT_MUTATION cdrewrite rule. This way is definitely more coherent though :)

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: I'll look into adding this! Also it turns out that `valid_token = GRAPHEMES.plus` isn't actually used in this function. The comparison actually uses 
```SENTENCE = (INITIAL_PUNCTUATION.ques + (GRAPHEMES.plus | NUMBERS.plus) + FINAL_PUNCTUATION.ques + (SPACE.plus + GRAPHEMES.plus).star + FINAL_PUNCTUATION.ques)```
which is a little bit fancier than just a sequence of graphemes, but I will for sure try to add the other valid tokens like email addresses etc.

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py:166]: I edited the right context of this rewrite rule and changed it from `union("[EOS]", "")` to `union("[EOS]", SPACE)`, so now it won't delete a word-initial apostrophe that hasn't been separated from the rest of the word (and it should now stay attached because of edits to the `SEPARATE_PUNCTUATION` rewrite rules).

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: The only things in this file that rely on the language are loading in the graphemes/punctuation and the language-specific rewrite rules. Perhaps we can have a main config that says the language we're using and/or which data sources we want to normalize and call it with something like `LANGUAGE = main_conf.LANGUAGE`?

One other possibility is turning it into a flag in `normalizer.py`, but I don't know how that would play with `normalizer_test.py`. Although, if the test file only uses the sample language and doesn't need language-specific anything, perhaps a flag could work if we can have it trickle into `normalizer_lib` when we run the binary.

File: normalizer_lib.py
Comment (sandyritchie) [inline normalizer_lib.py]: Instead of leaving this commented out, maybe make an alternative normalizer and add a test for it?

File: normalizer_lib.py
Comment (sandyritchie) [inline normalizer_lib.py]: If it's slow maybe just cut this altogether, and the function above.

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: I added a new class at the beginning of `normalizer_lib` to handle verbalizable tokens. It currently includes `EMAIL_ADDRESS`, `WEB_ADDRESS`, `TIME`, and  `FANCY_NUMBER` (which is for numbers with decimal/grouping separators like `.` and `,`). It also allows combinations of letters and numbers in tokens, so we now allow things like `mp3`.

We can follow-up on what other not-a-word tokens we want to include in a later PR.

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: That makes sense. Then we don't need to keep commenting different things in/out and can just call a different normalizer function if we want to filter out the whole sentence vs filtering out individual tokens.

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Agreed!

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: I went ahead and separated the normalizer function into two functions, one for token-based filtering and one for sentence-based filtering. In `normalizer_test.py` there are now separate tests for each one. In `normalizer.py`, if you give it a string to normalize it will call both functions so you can compare. If you don't give it a string (and it loads in a file instead), you can give it a flag telling it if you want to do token- or sentence-based filtering. It currently defaults to token-based.

One other thing to consider:  if we get a sentence that is e.g. 9 invalid tokens and 1 valid token, we might want to reject the whole sentence anyway. Maybe during the experiment phase we can explore different thresholds or methods for dealing with the fragments that result from replacing invalid tokens with `<UNK>`.

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py:68]: Could you add a comment here to explain what types of numbers this is accepting, as this is a bit difficult to parse :)

File: normalizer_lib.py
Comment (crewee) [inline normalizer_lib.py]: Remove the bit about pronunciation?

Comment (crewee) [reply normalizer_lib.py]: Also, we're missing 'step 3' haha :).  Just a nit.

**Review by crewee — APPROVED** [review]
LGTM after the last couple small nits :).

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py]: Done!

File: normalizer_lib.py
Comment (zupon) [reply normalizer_lib.py:68]: Added some comments explaining the ways these group numbers, with some examples.

**Pull request merged**
